### ISSUE TITLE
Calling close on a tcp socket should emit onDisconnect event

### ISSUE THREAD

**Issue by jpevarnek** [freedomjs/freedom-for-firefox#57]
Calling close on core.tcpsocket is expected to emit an onDisconnect event, that is not happening

**Comment by soycode** [comment]
In progress (#58) - with that change onConnection/onDisconnect events are being dispatched, but the order seems unpredictable. With some additional logging in freedom (https://github.com/freedomjs/freedom/compare/soycode-firefoxtests):

One run produces results like:

    console.log: SPOT 1
    console.log: SPOT 1.25
    console.log: ABOUT TO CONNECT
    console.log: IN CLIENTSOCKET CONNECT
    console.log: DONE WITH INITIAL CLIENTSOCKET CONNECT
    console.log: CLIENT FIRING ONDISCONNECT
    console.log: CLIENT FIRING ONCONNECTION
    console.log: SPOT 1.5
    console.warn: {}
    console.log: CLIENT FIRING ONDISCONNECT

While the next produces:

    console.log: SPOT 1
    console.log: SPOT 1.25
    console.log: ABOUT TO CONNECT
    console.log: IN CLIENTSOCKET CONNECT
    console.log: DONE WITH INITIAL CLIENTSOCKET CONNECT
    console.log: CLIENT FIRING ONCONNECTION
    console.log: SPOT 1.5
    console.warn: {}
    console.log: CLIENT FIRING ONDISCONNECT
    console.log: CLIENT FIRING ONDISCONNECT

In other words, still some async magic to tease out. I'd also be interested in seeing whatever pieces of uProxy are dependent on receiving these events, to have another test/basis of comparison.

**Comment by trevj** [comment]
Hey @soycode. Great to see this being fixed!

I think Simple SOCKS is a good place to test this out. It runs in Firefox and has clear logging statements for when TCP sockets are opened and closed. Right now, it's straightforward to compare the logs from running Simple SOCKS under Chrome and Firefox and to identify that TCP connection close events are not received in Firefox. Should be a good way to verify your fix.

**Comment by soycode** [comment]
Current status - the client socket does emit onConnection, but using its own dispatch, basically sending it to itself and not the corresponding server socket.

For instance, in the first integration test (https://github.com/freedomjs/freedom/blob/master/spec/providers/coreIntegration/tcpsocket.integration.src.js#L24-36) additional logging shows that the dispatch receives 'onConnection'. However in the actual onConnection/onDisconnect test (https://github.com/freedomjs/freedom/blob/master/spec/providers/coreIntegration/tcpsocket.integration.src.js#L123-133) the client is initialized with a jasmine spy rather than the same dispatch as the server, and so the dispatch never receives the event.

Fixing this basically means propagating messages from the client dispatch to the server it is connected to. This sounds right, and is probably what the Chrome provider is doing, but I wanted to document both to think through it and as a sanity check. Does this sound like expected/desired behavior? And should the client socket just pass everything on or only onConnection/onDisconnect events?

As an aside, in general it's been strange getting the Firefox sockets to behave the same as the Chrome ones, as the Chrome implementation is unified (server/client) while Firefox is a bit more split. I may revisit and do a more general refactor of Firefox sockets later, but the first priority is to fix this issue.

**Comment by willscott** [comment]
I'm not sure I understand the problem.  Why would the server auotmatically
learn that a different client has connected? what is the event flow we
expect to happen that isn't?

This is a client that has successfully connected to a server.  The server
listener should receive an 'accept' event, which results in a new client
socket spawned by the server. that new client would presumably already be
connected, and wouldn't need to emit an onConnect event.

**Comment by soycode** [comment]
Good point - in which case I may be testing the wrong thing in the test, and should potentially be doing the assertions for the client dispatch (i.e. Firefox behavior may be correct).

**Comment by soycode** [comment]
So I think this is probably what uProxy is expecting: https://github.com/uProxy/uproxy-networking/blob/c8ba97a44b59c2cd98917db65f4d33dce7d8b1fe/src/net/tcp.ts#L116

@jpevarnek, does this look like the right spot you expect something to happen? @willscott, are you suggesting that this could be on 'accept' instead of 'onConnection'?

**Comment by jpevarnek** [comment]
Three lines down ('onDisconnect') is the one the original bug concerned.

**Comment by soycode** [comment]
So I may have been approaching this incorrectly - I don't think we actually care about the client socket emitting anything, at least for this bug. We just want to make sure the server socket emits appropriate events on receiving clients. I'll work on a change for that, and maybe not address the client side at this point.

**Comment by soycode** [comment]
#59 may address this - tests still don't pass for reasons discussed earlier (basically the test client socket doesn't actually connect to the test server socket), but I'm hoping this will fix the uProxy case. I'll see if I can get it running in uproxy-networking, but other thoughts/eyes welcome.

**Comment by soycode** [comment]
Some progress - using the freedom-for-firefox.jsm built from #59 in the simple-socks-firefoxapp demo seems to give a few subtle differences in logging. In particular, https://github.com/uProxy/uproxy-networking/blob/55c2c4b9f5014f34664cb1c779c21be216dc8f73/src/net/tcp.ts#L363 is triggered - I'm not sure if that's actually desired or not, but it does indicate that the onDisconnectHandler is being called.

In other words, feedback/review welcome. I'll need to fix the integration test bit separately, but if it addresses the outstanding uProxy issue it's probably good to merge it in time for release. Thanks!

**Comment by trevj** [comment]
So, I've been trying this with Simple SOCKS.

For client sockets, there two cases:
 1. client sockets which use `connect`
 1. client sockets resulting from a connection to a server socket

I'm getting `onDisconnect` events for the first, but *not for the second*. This is easily verified by spinning up the echo server sample in uproxy-networking.

There's an additional problem with the argument passed to `onDisconnect` for which I've filed a separate issue:
https://github.com/freedomjs/freedom-for-firefox/issues/61

**Comment by soycode** [comment]
The server socket should dispatch the onDisconnect when it is explicitly closed (https://github.com/freedomjs/freedom-for-firefox/blob/master/providers/server_socket.js#L41) and when it receives `onStopListening` (https://github.com/freedomjs/freedom-for-firefox/blob/master/providers/server_socket.js#L17).

I'd hoped the latter would cover this situation - it refers to when the server 'stops for some reason' (https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIServerSocketListener#onStopListening()), but it looks like that reason doesn't include the client socket disconnecting from it (or I assume that's what you're referring to). I'll look for another event to cover it, let me know if you have other thoughts. Thanks!

**Comment by trevj** [comment]
Server sockets *definitely should not* dispatch `onDisconnect` when a client socket created as a result of a connection to the server socket terminates!

**Comment by soycode** [comment]
Ah alright, I don't think that *is* happening but good to know - so then the issue is that the client socket should dispatch but isn't when it was created from a connection to a server socket, and the behavior change is more likely to be on the client socket side than the server one.

**Comment by trevj** [comment]
Exactly.

**Comment by soycode** [comment]
So with the socket code in its current form I have an ugly but hopefully functional fix: https://github.com/freedomjs/freedom-for-firefox/commit/524c6ad6a50e1b78e1ad6f35acc2701544644b9e

Because servers/clients are all separated out I need to pass the dispatchEvent around to make this work. I think we've hit the "I should have refactored this already" point, but I'd still like to do a push of freedom-for-firefox this week that works with uProxy first, and then refactor next week (which will hopefully be easier if we have code that does what we want, however ugly it may be).

**Comment by soycode** [comment]
Revisiting - that fix has been merged but right now freedom-for-firefox is still pinned in uProxy. The pull request to unpin it is: https://github.com/uProxy/uproxy/pull/1545

When last tested (at this point about two weeks ago) it worked fine in Firefox 38, but may have had some issues in Firefox 37. I can't find a good way to keep a stable old install of past versions of Firefox (that don't want to auto-update all the time), and at this point I'm hoping FF37 is becoming less relevant regardless. Still would be good to try testing it again and if all is well merging and hopefully resolving this issue.

So, I'll try playing around with uProxy builds based on updated freedom-for-firefox, but also encourage you to check to see if it's resolved the issue from your perspective. Thanks!

**Comment by jpevarnek** [comment]
For future reference, you can grab old Firefox builds from https://ftp.mozilla.org/pub/mozilla.org/firefox/releases/

**Comment by soycode** [comment]
Yeah I've been able to install them but they rather insistently auto-update on first run. I'll dig and see if there's some about:config flag to stop them from doing that.

**Comment by trevj** [comment]
Can't you just dump the download in some random directory? I don't think it'll auto-update then. `cfx run` accepts an argument specifying a path to the binary.

**Comment by soycode** [comment]
Good idea - I'll give it a go.

**Comment by soycode** [comment]
I believe this is fixed.
